### PR TITLE
Update ExtensionResponseDTO to add custom analytics data to the message context

### PR DESCRIPTION
**Pull request by dimuthnc**
This PR enables sending custom analytics data from the ExtensionListner side to the APIM synapse context. This has been done using the 'CustomProperty' in the ExtensionListnerDTO. ( this was originally used to provide custom throttling data). 

Usecase: To publish Open Banking-specific custom data attributes as properties during the ELK data publishing by using the AnalyticsCustomDataProvider extension [1]. For this, the custom attributes need to be available in the message context.

Current Problem: The current extensions in the gateway flow do not provide direct access to set custom attributes to the message context and only have the ability to create a new ExtensionResponseDTO and return during the handler/filter flows in the API Gateway.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java [hunk 1]
```diff
@@ -264,10 +264,18 @@ private static boolean handleExtensionResponse(MessageContext messageContext,
         }
         // handle payload
         processResponsePayload(extensionResponseDTO, messageContext);
+        //Extract customProperty from DTO
+        Map<String, Object> customProperty = extensionResponseDTO.getCustomProperty();
+        //set Additional analytics Properties to the message context.
+        if(customProperty != null && customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS) != null ) {
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java:270]: Formating issue

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java [hunk 2]
```diff
@@ -264,10 +264,18 @@ private static boolean handleExtensionResponse(MessageContext messageContext,
...
+            Map<String, Object> customElkAttributes =
+                    (Map<String, Object>) customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS);
+            messageContext.setProperty(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS, customElkAttributes);
+            //removing additional analytics props as those are irrelevant to throttle stream
+            customProperty.remove(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS); }
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java:275]: Is this removal required for this fix?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java [hunk 3]
```diff
@@ -264,10 +264,18 @@ private static boolean handleExtensionResponse(MessageContext messageContext,
         }
         // handle payload
         processResponsePayload(extensionResponseDTO, messageContext);
+        //Extract customProperty from DTO
+        Map<String, Object> customProperty = extensionResponseDTO.getCustomProperty();
+        //set Additional analytics Properties to the message context.
+        if(customProperty != null && customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS) != null ) {
+            Map<String, Object> customElkAttributes =
+                    (Map<String, Object>) customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS);
+            messageContext.setProperty(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS, customElkAttributes);
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java:273]: Do we really need to do the casting to `Map<String, Object>` here as we are just setting it back to the `messageContext` as an object?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java [hunk 4]
```diff
@@ -264,10 +264,18 @@ private static boolean handleExtensionResponse(MessageContext messageContext,
         }
         // handle payload
         processResponsePayload(extensionResponseDTO, messageContext);
+        //Extract customProperty from DTO
+        Map<String, Object> customProperty = extensionResponseDTO.getCustomProperty();
+        //set Additional analytics Properties to the message context.
+        if(customProperty != null && customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS) != null ) {
+            Map<String, Object> customElkAttributes =
+                    (Map<String, Object>) customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS);
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java:272]: Do we really need to do the casting to `Map<String, Object>` here as we are just setting it back to the `messageContext` as an object?

**Review by lasanthaS — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java (hunk 4 above)
Comment (dimuthnc) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11772/commits/275e724ab7f475240e8ccb51f6d360e6c12ca298

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java [hunk 5]
```diff
@@ -267,9 +267,8 @@ private static boolean handleExtensionResponse(MessageContext messageContext,
         //Extract customProperty from DTO
         Map<String, Object> customProperty = extensionResponseDTO.getCustomProperty();
         //set Additional analytics Properties to the message context.
-        if(customProperty != null && customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS) != null ) {
-            Map<String, Object> customElkAttributes =
-                    (Map<String, Object>) customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS);
+        if(customProperty != null && customProperty.get(APIMgtGatewayConstants.ADDITIONAL_ANALYTICS_PROPS) != null) {
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java:270]: Need space after 'if'. @dimuthnc Please use code formatter plugin to format automatically.

Comment (dimuthnc) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/ext/listener/ExtensionListenerUtil.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11772/commits/970a6a2f8c745789ead1b8fd3bf4d0846cf21845

**Review by rmsamitha — APPROVED** [review]

**Review by lasanthaS — APPROVED** [review]
